velox/functions/sparksql/aggregates/ModeAggregate.cpp

```diff
@@ -0,0 +1,580 @@
+                16>(allocator)} {}
+
+  size_t size() const {
+    return values.size();
+  }
+
+  void addValue(
+      DecodedVector& decoded,
+      vector_size_t index,
+      HashStringAllocator* allocator) {
+    addValueWithCount(*decoded.base(), decoded.index(index), 1, allocator);
+  }
+
+  void addValueWithCount(
+      const BaseVector& keys,
+      vector_size_t index,
+      int64_t count,
+      HashStringAllocator* allocator) {
+    auto entry = serializedValues.append(keys, index, allocator);
+
+    auto it = values.find(entry);
+    if (it == values.end()) {
+      // New entry.
+      values[entry] += count;
+    } else {
+      // Existing entry.
+      serializedValues.removeLast(entry);
+
+      it->second += count;
+    }
+  }
+
+  void extractValues(
+      BaseVector& keys,
+      FlatVector<int64_t>& counts,
+      vector_size_t offset) {
+    auto index = offset;
+    for (const auto& [position, count] : values) {
+      AddressableNonNullValueList::read(position, keys, index);
+      counts.set(index, count);
+      ++index;
+    }
+  }
+
+  void extractValues(const VectorPtr& results, vector_size_t offset) {
+    auto iterator = std::max_element(
+        std::begin(values),
+        std::end(values),
+        [](const std::pair<AddressableNonNullValueList::Entry, int64_t>& p1,
+           const std::pair<AddressableNonNullValueList::Entry, int64_t>& p2) {
+          return p1.second < p2.second;
+        });
+    AddressableNonNullValueList::read(iterator->first, *results, offset);
+  }
+
+  void free(HashStringAllocator& allocator) {
+    serializedValues.free(allocator);
+    using TValues = decltype(values);
+    values.~TValues();
+  }
+};
+
+// Mode aggregate function for complex types.
+template <typename T>
+class ComplexTypeModeAggregate : public Aggregate {
+ public:
+  explicit ComplexTypeModeAggregate(TypePtr resultType, TypePtr inputType)
+      : Aggregate(std::move(resultType)), inputType_(std::move(inputType)) {}
+
+  int32_t accumulatorFixedWidthSize() const override {
+    return sizeof(ComplexTypeAccumulator);
+  }
+
+  bool isFixedSize() const override {
+    return false;
+  }
+
+  void extractValues(char** groups, int32_t numGroups, VectorPtr* result)
+      override {
+    VELOX_CHECK(*result);
+    (*result)->resize(numGroups);
+    uint64_t* rawNulls = nullptr;
+    if ((*result)->mayHaveNulls()) {
+      BufferPtr& nulls = (*result)->mutableNulls((*result)->size());
+      rawNulls = nulls->asMutable<uint64_t>();
+    }
+    for (auto i = 0; i < numGroups; ++i) {
+      char* group = groups[i];
+      auto* accumulator = value<ComplexTypeAccumulator>(group);
+      if (accumulator->size() == 0) {
+        (*result)->setNull(i, true);
+      } else {
+        if (rawNulls) {
+          bits::clearBit(rawNulls, i);
+        }
+        accumulator->extractValues(*result, i);
+      }
+    }
+  }
+
+  void extractAccumulators(char** groups, int32_t numGroups, VectorPtr* result)
```

Comment:
Any particular reason this function doesn't use Simple API? It would be nice to clarify in the PR description.

CC: @kagamiori

Reply:
`ComplexTypeAccumulator` needs inputType parameter, which doesn't align with `SimpleAggregateAdapter`'s intializer's signature.
https://github.com/facebookincubator/velox/blob/84428f9851ec0f0a04b5e75160b6400549366f8e/velox/exec/SimpleAggregateAdapter.h#L44C12-L44C34

Reply:
I see. Would you open an GitHub issue about this limitation in Simple Aggregate Function API. It would be nice to address. 

CC: @kgpai @kagamiori @pedroerp

Reply:
Opened https://github.com/facebookincubator/velox/issues/10728 to track this limitation.